### PR TITLE
Properly inform the user about software proposal problems

### PR DESCRIPTION
**Pull request by imobachgs**
## Problem

Recently, we added a *Software* section that informs the user about missing packages and the installation size. However, it has a few problems:

* When the network is not available (at this time, D-Installer uses online repos), it just reports that there are missing packages.
* While reading the repositories, the user has no feedback about the progress.

## Solution

* Display progress information while reading the repositories.
* Inform the user about problems reading the repositories.
* Add a mechanism to force refreshing the repositories.
* Reread the repositories if the network becomes available again.

## Screenshots

<details>
<summary>Progress information</summary>

Captura desde 2023-02-09 09-47-36
</details>

<details>
<summary>Report errors about unreachable repositories</summary>

Captura desde 2023-02-09 09-50-26
</details>

<details>
<summary>Do not inform about missing packages/patterns where there are no reachable repositories</summary>

Captura desde 2023-02-09 09-59-48
</details>

## To do

- [x] Move most of the proposal logic to a separate class ([DInstaller::Software::Manager](https://github.com/yast/d-installer/blob/master/service/lib/dinstaller/software/manager.rb#L41) contains a lot of logic).
- [x] Inform the user when a repository is not accessible.
- [x] Do not try to make a proposal if there are no repositories.
- [x] Display some progress in the corresponding section or, at least, explain what is happening.
- [x] Add a mechanism to allow refreshing the repositories (#388)
- [x] Refresh the repositories metadata when the network is available again

## Testing

- *Added a new unit test*
- *Tested manually*

### CODE REVIEW

File: service/lib/dinstaller/dbus/clients/network.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        def service_name
+          @service_name ||= "org.freedesktop.NetworkManager"
+        end
+
+        CONNECTED_NM_STATE = 70
+        private_constant :CONNECTED_NM_STATE
+
+        # Registers a callback to call when connectivity state changes
+        #
+        # The block receives a boolean argument which is true then the network
```
Comment (dgdavid) [inline service/lib/dinstaller/dbus/clients/network.rb:48]:
```suggestion
        # The block receives a boolean argument which is true when the network
```

File: service/lib/dinstaller/software/proposal.rb [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+      # Extracts the warning messages from the proposal result
+      #
+      # @param proposal_result [Hash] Proposal result; it might contain a "warning" key with warning
+      #   messages.
+      def warning_messages(proposal_result)
+        return [] unless proposal_result["warning_level"] == :blocker
+
+        proposal_result["warning"]
+          .split("<br>")
+          .grep_v(/Please manually select .*/) # FIXME: it depends on the language
```
Comment (dgdavid) [inline service/lib/dinstaller/software/proposal.rb:193]: Should we create a card for this FIXME?

File: service/test/dinstaller/software/proposal_test.rb [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+          [DInstaller::ValidationError.new("Could not install...")]
+        )
+      end
+    end
+
+    context "when solver errors are reported" do
+      let(:last_error) { "Solving errors..." }
+      let(:solve_errors) { 5 }
+
+      it "returns something" do
```
Comment (dgdavid) [inline service/test/dinstaller/software/proposal_test.rb:103]: NP: I'd say "registers them as validation errors" or so. In fact, you're not testing the return value of #calculate here.

File: service/test/dinstaller/software/proposal_test.rb [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+          [
+            DInstaller::ValidationError.new("Solving errors..."),
+            DInstaller::ValidationError.new("Found 5 dependency issues.")
+          ]
+        )
+      end
+    end
+  end
+
+  describe "#validate" do
```
Comment (dgdavid) [inline service/test/dinstaller/software/proposal_test.rb:115]: Do you mean #errors?

File: service/test/dinstaller/software/proposal_test.rb [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    context "when no proposal was calculated" do
+      it "returns an empty array" do
+        expect(subject.errors).to eq([])
+      end
+    end
+
+    context "when a blocking was reported" do
+      it "returns a validation error for the package/pattern"
```
Comment (dgdavid) [inline service/test/dinstaller/software/proposal_test.rb:123]: Leftover? Or just incomplete?

File: service/test/dinstaller/software/proposal_test.rb [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    end
+  end
+
+  describe "#set_resolvables" do
+    it "adds the list of packages/patterns to the proposal"
```
Comment (dgdavid) [inline service/test/dinstaller/software/proposal_test.rb:128]: Same than above, leftover or incomplete

File: web/src/components/core/FileViewer.test.jsx [hunk 7]
```diff
@@ -28,19 +28,19 @@ import cockpit from "../../lib/cockpit";
...
     read: readFn
-  }
+  };
 });
 
 cockpit.file.mockImplementation(fileFn);
 
 // testing data
-const file_name = "/testfile"
+const file_name = "/testfile";
```
Comment (dgdavid) [inline web/src/components/core/FileViewer.test.jsx:43]: Just a reminder for myself: The same than https://github.com/yast/d-installer/pull/418#pullrequestreview-1289770888

File: web/src/components/core/FileViewer.jsx [hunk 8]
```diff
@@ -62,16 +62,16 @@ export default function FileViewer({ file, title, onCloseCallback }) {
...
+      {(content === null || error) &&
         <Alert
           isInline
           isPlain
           variant="warning"
           title="Cannot read the file"
         >
           {error}
-        </Alert> }
+        </Alert>}
```
Comment (dgdavid) [inline web/src/components/core/FileViewer.jsx:74]: I guess this change is just for making ESLint happy, as it's partially done in https://github.com/yast/d-installer/pull/418#pullrequestreview-1289770888

File: web/src/components/core/ProgressText.jsx [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+ * @param {object} props
+ * @param {string} [props.message] Progress message
+ * @param {number} [props.current] Current step
+ * @param {number} [props.total] Number of steps
+ */
+export default function ProgressText({ message, current, total }) {
+  const text = (current === 0) ? message : `${message} (${current}/${total})`;
+  return (
+    <div className="split">
+      <Spinner size="md" aria-label="Progress indicator" />
```
Comment (dgdavid) [inline web/src/components/core/ProgressText.jsx:41]: I know we were talking about this, but maybe worth it to test how it looks with `<Icon name="loading" size="24" />` (in fact, I already did it) just for the sake of using just one "spinner" across the app. Although, as we discussed, we can distinguish between "Loading page" and "Loading state" and keep the consistence accordingly.

| Spinner | Loading Icon |
|:-:|:-:|
| spinner | loading |

File: web/src/components/network/WifiNetworksList.test.jsx [hunk 10]
```diff
@@ -21,7 +21,7 @@
 
 import React from "react";
 
-import { screen, waitFor } from "@testing-library/react";
+import { screen } from "@testing-library/react";
```
Comment (dgdavid) [inline web/src/components/network/WifiNetworksList.test.jsx:24]: Note: not related with the PR, just pleasing ESLint

**Review by dgdavid — COMMENTED** [review]
Actually it looks good, just a few comments.

File: service/test/dinstaller/software/proposal_test.rb (hunk 4 above)
Comment (imobachgs) [reply service/test/dinstaller/software/proposal_test.rb:115]: Actually, none of them :-)

File: service/test/dinstaller/software/proposal_test.rb (hunk 6 above)
Comment (imobachgs) [reply service/test/dinstaller/software/proposal_test.rb:128]: I forgot to check incomplete tests 😞

File: service/lib/dinstaller/software/proposal.rb (hunk 2 above)
Comment (imobachgs) [reply service/lib/dinstaller/software/proposal.rb:193]: Yes. Actually, we need to go through all the FIXMEs and TODOs...

**Review by dgdavid — APPROVED** [review]
LGTM :+1: , thank you!

**Review by dgdavid — APPROVED** [review]
LGTM

**Pull request merged**